mathesar_ui/src/sections/table-view/header/HeaderCellDropdownTypeOptions.svelte

```diff
@@ -0,0 +1,225 @@
+      (mathesarType) => pair(
+        mathesarType.identifier,
+        getValidDbTypeTargetsForColumnAndMathesarType(column, mathesarType),
+      ),
+    );
+    return new Map(pairs);
+  }
+
+  function mathesarTypeHasAtLeastOneValidDbTypeTarget(
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    validDbTypeTargetsPerMathesarType: DbTypeTargetsPerMathesarType,
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    mathesarType: MathesarType,
+  ) {
+    // eslint-disable-next-line operator-linebreak
+    const validDbTypeTargets =
+      validDbTypeTargetsPerMathesarType.get(mathesarType.identifier);
+    const atLeastOne = notEmpty(validDbTypeTargets);
+    return atLeastOne;
+  }
+
+  function patchColumnToMathesarType(
+    // as returned by createEventDispatcher()
+    dispatch: (x: unknown) => void,
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    tableId: number,
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    columnId: number,
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    validDbTypeTargetsPerMathesarType: DbTypeTargetsPerMathesarType,
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    mathesarType: MathesarType,
+  ): void {
+    if (validDbTypeTargetsPerMathesarType) {
+      const newDbType = choosePreferredDbTypeTarget(mathesarType);
+      const reloadTable = () => dispatch('reload');
+      void patchColumnType(tableId, columnId, newDbType)
+        .then(reloadTable);
+    }
+  }
+</script>
+
+<!--
+  Svelte linter cannot differentiate between static/pure and stateful functions
+so it warns when using module context functions in reactive blocks.
+
+  The ability to disable this specific warning will be available when svelte is
+updated to version 3.39.0 or above. The svelte-ignore directive below will then
+take effect.
+
+  https://github.com/sveltejs/svelte/issues/5954
+  https://github.com/sveltejs/svelte/pull/6504
+  -->
+
+<!-- svelte-ignore module-script-reactive-declaration -->
+<script lang="ts">
+  import type { DbType, MathesarType } from '@mathesar/stores/databases';
+  import { getMathesarTypeIcon } from '@mathesar/stores/databases';
+  import type { TableColumn } from '@mathesar/stores/tableData';
+  import { Button } from '@mathesar-components';
+  import { isDefined } from '@mathesar/utils/language';
+
+  export let mathesarTypes: MathesarType[];
+  export let tableId: number;
+  export let column: TableColumn;
+  export let isOpen: boolean;
+  export let isDataTypeOptionsOpen: boolean;
+
+  $: columnId = column.index;
+
+  let validDbTypeTargetsPerMathesarType: DbTypeTargetsPerMathesarType | undefined;
+  $: {
+    if (mathesarTypes) {
+      // eslint-disable-next-line operator-linebreak
+      validDbTypeTargetsPerMathesarType =
+        getValidDbTypeTargetsPerMathesarType(column, mathesarTypes);
+    }
+  }
+
+  const implementedMathesarTypes: MathesarType['identifier'][] = ['number', 'text'];
```

Comment:
We need to list all types on the frontend, even the ones that are not ready in the backend. Frontend should not contain hardcode displayable types.

    >  a MathesarType array on CellHeaderDropdownTypeOptions will have to be updated to include the newly supported types

We did cover this  on #498, but I guess we did not go into detail or if I did not explain it well. Having the code makes it easier to specify what I had in mind.

The changes needed here is relatively simple. We would only need to remove the check based on the hardcoded array. Everything will be shown.

We will however hide/disable the invalid options based on column.valid_target_types.

### The only allowed temporary hardcoding, until the backend is ready:
1. default_db_target_type for a mathesar_type: We can hardcode text->'VARCHAR' and number->'NUMERIC'. This seems to be implemented in `choosePreferredDbTypeTarget` method in `databases.ts` store. For other types, instead of throwing error, we can choose the first entry from the mathesar_type -> db_type[] map.
2. Icons for mathesar_type: This is also already implemented.

Reply:
> We will however hide/disable the invalid options based on column.valid_target_types.

If that isn't clear, it's already being done by looking at the intersection of `Column.valid_target_types` and `MathesarType.db_types`.

Reply:
> We will however hide/disable the invalid options based on column.valid_target_types.

@dmos62 I did look at the intersection, it works as expected. Since I reiterated the expected behaviour in the above comment, I mentioned this. I should have clarified that it already exists.

Reply:
No problem at all, just wanted to make sure we're clear.